lib/cache/collections.go

```diff
@@ -2686,3 +2691,95 @@ func (s *kubeCluster) processEvent(ctx context.Context, event types.Event) error
 func (s *kubeCluster) watchKind() types.WatchKind {
 	return s.watch
 }
+
+type samlIDPServiceProviders struct {
+	*Cache
+	watch types.WatchKind
+}
+
+func (s *samlIDPServiceProviders) erase(ctx context.Context) error {
+	if err := s.samlIdpServiceProvidersCache.DeleteAllSAMLIdPServiceProviders(ctx); err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func (s *samlIDPServiceProviders) fetch(ctx context.Context) (apply func(ctx context.Context) error, err error) {
+	resources := make([]types.SAMLIdPServiceProvider, 0)
+
+	nextKey := ""
+	for {
+		var samlProviders []types.SAMLIdPServiceProvider
+		var err error
+		samlProviders, nextKey, err = s.SAMLIdPServiceProviders.ListSAMLIdPServiceProviders(ctx, 200, nextKey)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		resources = append(resources, samlProviders...)
+		if nextKey == "" {
+			break
+		}
+	}
+
+	return func(ctx context.Context) error {
+		if err := s.erase(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+
+		for _, resource := range resources {
+			if _, err := s.samlIdpServiceProvidersCache.GetSAMLIdPServiceProvider(ctx, resource.GetName()); err != nil {
+				if trace.IsNotFound(err) {
+					err = s.samlIdpServiceProvidersCache.CreateSAMLIdPServiceProvider(ctx, resource)
+					if err != nil {
+						return trace.Wrap(err)
+					}
+				}
+			}
```

Comment:
```suggestion
			if _, err := s.samlIdpServiceProvidersCache.GetSAMLIdPServiceProvider(ctx, resource.GetName()); trace.IsNotFound(err) {
					err = s.samlIdpServiceProvidersCache.CreateSAMLIdPServiceProvider(ctx, resource)
					if err != nil {
						return trace.Wrap(err)
					}
				
			}else if err!=nil{
			
			return trace.Wrap(err)
			}
```

Should we upsert when we create the resource or should we just create and return?

Reply:
Ahh whoops, you're right! I'll return after the create.